display_list/display_list_color_filter.h

```diff
@@ -0,0 +1,240 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_DISPLAY_LIST_DISPLAY_LIST_COLOR_FILTER_H_
+#define FLUTTER_DISPLAY_LIST_DISPLAY_LIST_COLOR_FILTER_H_
+
+#include "flutter/display_list/types.h"
+#include "flutter/fml/logging.h"
+
+namespace flutter {
+
+class DlBlendColorFilter;
+class DlMatrixColorFilter;
+class DlSrgbToLinearGammaColorFilter;
+class DlLinearToSrgbGammaColorFilter;
+class DlUnknownColorFilter;
+
+class DlColorFilter {
+ public:
+  enum Type {
+    kBlend,
+    kMatrix,
+    kSrgbToLinearGamma,
+    kLinearToSrgbGamma,
+    kUnknown
+  };
+
+  static std::shared_ptr<DlColorFilter> From(SkColorFilter* sk_filter);
+  static std::shared_ptr<DlColorFilter> From(sk_sp<SkColorFilter> sk_filter) {
+    return From(sk_filter.get());
+  }
+
+  virtual Type type() const = 0;
+  virtual size_t size() const = 0;
+  virtual bool modifies_transparent_black() const = 0;
+
+  static std::shared_ptr<DlColorFilter> Shared(const DlColorFilter* filter) {
+    return filter == nullptr ? nullptr : filter->shared();
+  }
+  virtual std::shared_ptr<DlColorFilter> shared() const = 0;
+  virtual sk_sp<SkColorFilter> sk_filter() const = 0;
+
+  // asSrgb<->Linear and asUnknown are not needed because they
+  // have no properties to query. Their type fully specifies their
+  // operation or can be accessed via the common sk_filter() method.
+  virtual const DlBlendColorFilter* asBlend() const { return nullptr; }
+  virtual const DlMatrixColorFilter* asMatrix() const { return nullptr; }
+
+  bool operator==(DlColorFilter const& other) const {
+    return type() == other.type() && equals_(other);
+  }
+  bool operator!=(DlColorFilter const& other) const {
+    return !(*this == other);
+  }
+
+  virtual ~DlColorFilter() = default;
+
+ protected:
+  virtual bool equals_(DlColorFilter const& other) const = 0;
+};
+
+class DlBlendColorFilter final : public DlColorFilter {
+ public:
+  DlBlendColorFilter(SkColor color, SkBlendMode mode)
+      : color_(color), mode_(mode) {}
+  DlBlendColorFilter(const DlBlendColorFilter& filter)
+      : DlBlendColorFilter(filter.color_, filter.mode_) {}
+  DlBlendColorFilter(const DlBlendColorFilter* filter)
+      : DlBlendColorFilter(filter->color_, filter->mode_) {}
+
+  Type type() const override { return kBlend; }
+  size_t size() const override { return sizeof(*this); }
+  bool modifies_transparent_black() const override {
+    // Look at blend and color to make a faster determination?
+    return sk_filter()->filterColor(SK_ColorTRANSPARENT) != SK_ColorTRANSPARENT;
+  }
+
+  std::shared_ptr<DlColorFilter> shared() const override {
+    return std::make_shared<DlBlendColorFilter>(this);
```

Comment:
Why isn't this unsafe? Each call to this will create new shared pointers that don't know about each other. One of them could end up dangling.

Reply:
make_shared<T>(...) calls the constructor on T that matches the args. It returns a shared_ptr to a brand new instance of the DlColorFilter (DlBlendColorFilter in that particular case).

`shared()->get() != this`
`*shared()->get() == *this`

Perhaps I should add tests for that first condition to the unittests?

Reply:
My concept was that the DlFoo objects would be lightweight, immutable, and copyable without loosing any functionality. This function takes advantage of those design constraints.

Reply:
The only unsafe (perhaps surprising) aspect of this method is that the address in the shared_ptr will not be the address of the original. That might be unexpected, but I don't think this is unsafe wrt memory management and since the objects are immutable, it shouldn't be unsafe with respect to sharing modifications.

Reply:
I added those unit tests to verify that the shared ptr returned is distinct from the original.